pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/ClassStats.java

```diff
@@ -159,14 +159,14 @@
      *
      * @return The result of the computation, or {@code Double.NaN} if it couldn't be performed
      */
-    /* default */ double computeWithResultOption(OperationMetricKey key, ASTAnyTypeDeclaration node,
+    /* default */ double computeWithResultOption(MetricKey<OperationMetric> key, ASTAnyTypeDeclaration node,
                                                  boolean force, MetricVersion version, ResultOption option) {
 
         List<ASTMethodOrConstructorDeclaration> ops = findOperations(node, false);
 
         List<Double> values = new ArrayList<>();
         for (ASTMethodOrConstructorDeclaration op : ops) {
-            if (key.supports(op)) {
+            if (key.getCalculator().supports(op)) {
```

Comment:
Hmm I think that through generifying the interfaces we could achieve an even more specific result. If you generify `Metric` into `Metric<N>`, where `N` stands for the type of node the metric can be computed on, that would look like so:

```java
interface Metric<N extends Node> {
  boolean supports(N node);
  double computeFor(N node, MetricVersion v);
}

interface ClassMetric extends Metric<ASTAnyTypeDeclaration> {}
interface OperationMetric extends Metric<ASTMethodOrConstructorDeclaration> {}

interface MetricKey<N> {
  boolean supports(N node);
  Metric<N> getCalculator();
  String name();
}
```
The downside is that no class could implement ClassMetric or OperationMetric at the same time, as that would make Metric be inherited with different type arguments. Though I think it could be good to separate ClassMetrics' implementations from their OperationMetric, as Metric objects represent behaviour (they could nearly be seen as a lambda if it wasn't for `support`). The fact that concrete metric implement both interfaces as of now seems contrary to that design, as it mixes conceptually unrelated behaviours.

Would you let me work on that for the PR after the next ? I already have a few commits after these ones in my local repo